File: MessageServer/MessageServer.py
#!/usr/bin/env python3
import sys
import os
import getopt

srcroot=os.path.abspath(os.path.dirname(os.path.abspath(__file__))+"/..")
sys.path.append(srcroot+"/MsgApp")

from PyQt5 import QtCore, QtGui, QtWidgets, QtNetwork

from TcpServer import *
from WebSocketServer import *

from Messaging import Messaging

class MessageServer(QtWidgets.QMainWindow):
    def __init__(self, argv):
        QtWidgets.QMainWindow.__init__(self)
        
        self.settings = QtCore.QSettings("MsgTools", "MessageServer")
        self.logFile = None
        
        srcroot=os.path.abspath(os.path.dirname(os.path.abspath(__file__))+"/..")
        msgdir = srcroot+"/../obj/CodeGenerator/Python/"
        self.msgLib = Messaging(msgdir, 0, "NetworkHeader")
        self.connectClass = Messaging.MsgClassFromName["Network.Connect"]
        self.subscriptionListClass = Messaging.MsgClassFromName["Network.SubscriptionList"]
        self.maskedSubscriptionClass = Messaging.MsgClassFromName["Network.MaskedSubscription"]

        self.clients = {}

        self.initializeGui()

        self.tcpServer = TcpServer()
        self.tcpServer.statusUpdate.connect(self.onStatusUpdate)
        self.tcpServer.newConnection.connect(self.onNewConnection)

        self.wsServer = WebSocketServer()
        self.wsServer.statusUpdate.connect(self.onStatusUpdate)
        self.wsServer.newConnection.connect(self.onNewConnection)

        options = ['serial=', 'bluetoothSPP=', 'plugin=']
        self.optlist, args = getopt.getopt(sys.argv[1:], '', options)
        for opt in self.optlist:
            if opt[0] == '--serial':
                from SerialHeader import SerialHeader
                from SerialPlugin import SerialConnection
                serialPortName = opt[1]
                self.serialPort = SerialConnection(SerialHeader, serialPortName)
                self.serialPort.statusUpdate.connect(self.onStatusUpdate)
                self.onNewConnection(self.serialPort)
                self.serialPort.start()
            elif opt[0] == '--bluetoothSPP':
                from BluetoothHeader import BluetoothHeader
                from SerialPlugin import SerialConnection
                bluetoothPortName = opt[1]
                self.bluetoothPort = SerialConnection(BluetoothHeader, bluetoothPortName)
                self.bluetoothPort.statusUpdate.connect(self.onStatusUpdate)
                self.onNewConnection(self.bluetoothPort)
                self.bluetoothPort.start()
            elif opt[0] == '--plugin':
                filename = opt[1]
                moduleName = os.path.splitext(os.path.basename(filename))[0]
                if Messaging.debug:
                    print("loading module ", filename, "as",moduleName)
                self.plugin = imp.load_source(filename.replace("/", "_"), filename)

                self.pluginPort = self.plugin.PluginConnection()
                self.pluginPort.statusUpdate.connect(self.onStatusUpdate)
                self.pluginPort.newConnection.connect(self.onNewConnection)
                self.pluginPort.start()

        self.tcpServer.start()
        self.wsServer.start()
        name = self.tcpServer.serverInfo() + "(TCP) and " + str(self.wsServer.portNumber) + "(WebSocket)"
        self.statusBar().addPermanentWidget(QtWidgets.QLabel(name))

    def initializeGui(self):
        # Layout
        vbox = QtWidgets.QVBoxLayout()
        self.grid = QtWidgets.QGridLayout()
        vbox.addLayout(self.grid)

        # Components
        self.logButton = QtWidgets.QPushButton("Start Logging")
        self.logButton.pressed.connect(self.onLogButtonClicked)
        vbox.addWidget(self.logButton)

        self.statusBox = QtWidgets.QPlainTextEdit()
        vbox.addWidget(self.statusBox)

        # Central Widget (QMainWindow limitation)
        centralWidget = QtWidgets.QWidget()
        centralWidget.setLayout(vbox)
        self.setCentralWidget(centralWidget)

        # Main Window Stuff
        self.setWindowTitle("MessageServer 0.1")
        self.setGeometry(300, 100, 800, 400)
        self.statusBar()
    
    def onLogButtonClicked(self):
        if self.logFile != None:
            self.logFile.close()
            self.logFile = None
            self.logButton.setText("Start Logging")
        else:
            currentDateTime = QtCore.QDateTime.currentDateTime()
            defaultFilename = currentDateTime.toString("yyyyMMdd-hhmmss") + ".log"
            logFileName, _ = QtWidgets.QFileDialog.getSaveFileName(self, "Save File", self.settings.value("logging/filename", ".")+"/"+defaultFilename, "Log Files (*.log)")
            # if they hit cancel, don't do anything
            if not logFileName:
                return
            self.logFile = QtCore.QFile(logFileName)
            self.logFile.open(QtCore.QIODevice.Append)
            fileInfo = QtCore.QFileInfo(logFileName)
            self.settings.setValue("logging/filename", fileInfo.dir().absolutePath())
            self.logButton.setText("Stop " + fileInfo.fileName())

    def onStatusUpdate(self, message):
        self.statusBox.appendPlainText(message)

    def onNewConnection(self, newConnection):
        self.onStatusUpdate("adding connection[" + newConnection.name+"]")
        self.clients[newConnection] = newConnection
        newConnection.messagereceived.connect(self.onMessageReceived)
        newConnection.disconnected.connect(self.onConnectionDied)
        clientRow = self.grid.rowCount()
        i = 0
        while(1):
            widget = newConnection.widget(i)
            if widget == None:
                break
            self.grid.addWidget(widget, clientRow, i)
            i+=1

    def onConnectionDied(self, connection):
        self.onStatusUpdate("removing connection[" + connection.name+"]")
        i = 0
        while(1):
            widget = connection.widget(i)
            if widget == None:
                break
            self.grid.removeWidget(widget)
            widget.deleteLater()
            i+=1
        if connection in self.clients:
            connection.deleteLater()
            del self.clients[connection]
        else:
            self.onStatusUpdate("cnx not in list!")

    def onMessageReceived(self, message):
        c = self.sender()
        # check for name, subscription, etc.
        if Messaging.hdr.GetMessageID(message) == self.connectClass.ID:
            c.name = self.connectClass.GetName(message)
            c.statusLabel.setText(c.name)
        elif Messaging.hdr.GetMessageID(message) == self.subscriptionListClass.ID:
            c.subscriptions = {}
            for idx in range(0,self.subscriptionListClass.GetIDs.count):
                id = self.subscriptionListClass.GetIDs(message, idx)
                if id != 0:
                    c.subscriptions[id] = id
            self.onStatusUpdate("updating subscription for "+c.name+" to " + ', '.join(hex(x) for x in c.subscriptions.keys()))
        elif Messaging.hdr.GetMessageID(message) == self.maskedSubscriptionClass.ID:
            c.subMask = self.maskedSubscriptionClass.GetMask(message)
            c.subValue = self.maskedSubscriptionClass.GetValue(message)
            self.onStatusUpdate("updating subscription for "+c.name+" to id & " + hex(c.subMask) + " == " + hex(c.subValue))
        else:
            #write to log, if log is open
            if self.logFile != None:
                self.logFile.write(message)
        for client in self.clients.values():
            if client != c:
                id = Messaging.hdr.GetMessageID(message)
                if id in client.subscriptions or (id & client.subMask == client.subValue):
                    client.sendMsg(message)

if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)

    msgServer = MessageServer(sys.argv)
    msgServer.show()

    sys.exit(app.exec_())

File: MessageServer/MessageServer.cpp
#include <QtWidgets/QGroupBox>
#include <QtWidgets/QFileDialog>
#include <QtWidgets/QStatusBar>
#include <QtWidgets/QVBoxLayout>
#include <QList>
#include <QtNetwork>
#include <qapplication.h>

#include <stdlib.h>

#include "MessageServer.h"
#include "Client.h"
#include "ServerInterface.h"

MessageServer::MessageServer(int /*argc*/, char */*argv*/[])
: QMainWindow(),
  _logFile(0),
  _settings("MsgTools", "MessageServer")
{
    _statusBox = new QPlainTextEdit();
    _statusBox->setMaximumBlockCount(10000);
    qInstallMessageHandler(MessageServer::redirectDebugOutput);

    _tcpServer = new QTcpServer(this);
    if (!_tcpServer->listen(QHostAddress::Any, 5678))
    {
        qWarning() << "Unable to start the server: " << _tcpServer->errorString() << endl;
        exit(1);
        return;
    }

    /** Show IP address and port number in status bar */
    QString name = "";
    foreach (const QHostAddress &address, QNetworkInterface::allAddresses())
    {
        if (address.protocol() == QAbstractSocket::IPv4Protocol &&
            address != QHostAddress(QHostAddress::LocalHost))
        {
            QString addrStr = address.toString();
            // ignore VM/VPN stuff (special ubuntu address, and anything that ends in 1)
            if(addrStr.section( ".",-1,-1 ) != "1" && addrStr != "192.168.122.1")
               name += address.toString() + "/";
        }
    }
    name.remove(-1, 1);
    name += QString(" : %1").arg(_tcpServer->serverPort());
    statusBar()->addPermanentWidget(new QLabel(name));

    connect(_tcpServer, &QTcpServer::newConnection, this, &MessageServer::GotANewClient);

    QVBoxLayout* vbox = new QVBoxLayout();
    _layout = new QGridLayout();
    
    QGroupBox* box = new QGroupBox;
    box->setLayout(vbox);
    setCentralWidget(box);

    _logButton = new QPushButton("Start Logging");
    connect(_logButton, &QPushButton::clicked, this, &MessageServer::LogButtonClicked);
    vbox->addWidget(_logButton);

    _loadPluginButton = new QPushButton("Load Plugin");
    connect(_loadPluginButton, &QPushButton::clicked, this, &MessageServer::LoadPluginButton);
    vbox->addWidget(_loadPluginButton);

    vbox->addLayout(_layout);
    vbox->addWidget(_statusBox);
}
void MessageServer::LogButtonClicked()
{
    if(_logFile)
    {
        _logFile->close();
        _logFile.clear();
        _logButton->setText("Start Logging");
    }
    else
    {
        QDate d = QDate::currentDate();
        QTime t = QTime::currentTime();
        QString defaultFilename = d.toString("yyyyMMdd") + "-" + t.toString("hhmmss") + ".log";
        QString logFileName =
            QFileDialog::getSaveFileName
            (this, tr("Save File"),
             _settings.value("logging/filename", ".").toString()+"/"+defaultFilename,
             tr("Log Files (*.log)"));
        // if they hit cancel, don't do anything
        if(logFileName.isNull())
            return;
        _logFile = QSharedPointer<QFile>(new QFile(logFileName));
        _logFile->open(QIODevice::Append);
        QFileInfo fileInfo(logFileName);
        _settings.setValue("logging/filename", fileInfo.dir().absolutePath());
        _logButton->setText(QString("Stop %1").arg(fileInfo.fileName()));
    }
}

void MessageServer::LoadPluginButton()
{
    QString pluginFileName =
        QFileDialog::getOpenFileName
        (this, tr("Open Plugin"),
         _settings.value("plugin/filename", QCoreApplication::applicationDirPath()).toString(),
         tr("Plugin Files (*.plugin)"));
    // if they hit cancel, don't do anything
    if(pluginFileName.isNull())
        return;
    _settings.setValue("plugin/filename", pluginFileName);
    LoadPlugin(pluginFileName);
}

void MessageServer::MessageSlot(QSharedPointer<Message> msg)
{
    if(_logFile)
    {
        _logFile->write((const char*)msg->RawBuffer(), Message::HeaderSize());
        _logFile->write((const char*)msg->GetDataPtr(), msg->GetDataLength());
    }
}

void MessageServer::AddNewClient(ServerPort* serverPort)
{
    // connect the new client to all the existing clients
    for(int i=0; i<_clients.count(); i++)
    {
        connect(serverPort,     &ServerPort::MsgSignal, _clients.at(i), &ServerPort::MessageSlot);
        connect(_clients.at(i), &ServerPort::MsgSignal, serverPort,     &ServerPort::MessageSlot);
    }

    // add the new client to the list
    _clients.append(serverPort);
    connect(serverPort, &ServerPort::MsgSignal, this, &MessageServer::MessageSlot);
    int clientRow = _layout->rowCount();
    for(int i=0; ; i++)
    {
        QWidget* widget = serverPort->widget(i);
        if(!widget)
            break;
        _layout->addWidget(widget, clientRow, i);
    }

    connect(serverPort, &ServerPort::disconnected, this, &MessageServer::ClientDied);
}

void MessageServer::GotANewClient()
{
    QTcpSocket *clientSocket = _tcpServer->nextPendingConnection();
    Client* clientConnection = new Client(clientSocket);

    AddNewClient(clientConnection);
}

// remove dead clients from the list of clients
void MessageServer::ClientDied()
{
    QObject* s = QObject::sender();
    /** \note this used to be qobject_cast<>, which failed for an object that was a subclass of
     * ServerPort defined in the SerialPlugin.  Changing to dynamic_cast<> made it work again. */
    ServerPort* serverPort = dynamic_cast<ServerPort*>(s);
    if(serverPort)
    {
        if(_clients.removeOne(serverPort))
        {
            qDebug() << "Removed client " << serverPort->Name() << endl;
            statusBar()->showMessage(QString("Removed %1").arg(serverPort->Name()), 1000);
            for(int i=0; ; i++)
            {
                QWidget* widget = serverPort->widget(i);
                if(!widget)
                    break;
                _layout->removeWidget(widget);
            }
            delete serverPort;
        }
        else
        {
            qCritical() << "Failed to remove client " << endl; // << dbConn->Name() << endl;
        }
    }
    else
    {
        qCritical() << "got ClientDied from sender not a ServerPort " << endl;
    }
}

void MessageServer::LoadPlugin(QString fileName)
{
    // resolve the correct file name for libraries on the given platform
    QLibrary lib(fileName);
    if(!lib.load())
    {
        qWarning() << fileName << " cannot be loaded by QLibrary (" << lib.errorString() << ")" << endl;
        return;
    }
    fileName = lib.fileName();

    QPluginLoader loader(fileName);
    QObject *plugin = loader.instance();

    if (plugin)
    {
        ServerInterface* dbPlugin = qobject_cast<ServerInterface*>(plugin);
        if (dbPlugin)
        {
            ServerPort* conn = dbPlugin->CreateConnection();
            AddNewClient(conn);
        }
        else
        {
            qWarning() << fileName << " is not a ServerInterface" << endl;
        }
    }
    else
    {
        qWarning() << fileName << " cannot be loaded by QPluginLoader (" << loader.errorString() << ")" << endl;
    }
}

void MessageServer::redirectDebugOutput(QtMsgType type, const QMessageLogContext& /*context*/, const QString &msg)
{
    QString output = msg;// + "("+context.file + ":" + context.line + ", " + context.function + ")";
    switch (type)
    {
    case QtDebugMsg:
        Instance()->_statusBox->appendPlainText(output);
        break;
    case QtWarningMsg:
        Instance()->_statusBox->appendPlainText("Warning: " + output);
        break;
    case QtCriticalMsg:
        Instance()->_statusBox->appendPlainText("Critical: " + output);
        break;
    case QtFatalMsg:
        fprintf(stderr, "Fatal: %s", output.toUtf8().constData());
        abort();
    case QtInfoMsg:
        break;
    }
}

MessageServer* MessageServer::Instance(int argc, char** argv)
{
    static MessageServer* instance = new MessageServer(argc, argv);
    return instance;
}
